src/typed-ast/expr.h

```diff
@@ -0,0 +1,637 @@
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("bool-literal").field(value).finish();
+    }
+  };
+
+  class StringLiteral final : public Expr
+  {
+    std::string value;
+
+  public:
+    explicit StringLiteral(SourceLocation location, std::string value)
+    : Expr(Kind::StringLiteral, location), value(value)
+    {}
+
+    const std::string& getValue() const
+    {
+      return value;
+    }
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::StringLiteral;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("string-literal").field(value).finish();
+    }
+  };
+
+  class InterpolateExpr final : public Expr
+  {
+    std::vector<ExprPtr> elements;
+
+  public:
+    explicit InterpolateExpr(
+      SourceLocation location, std::vector<ExprPtr> elements)
+    : Expr(Kind::Interpolate, location), elements(std::move(elements))
+    {}
+
+    std::vector<ExprPtr>& getElements()
+    {
+      return elements;
+    }
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Interpolate;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("seq").field(elements).finish();
+    }
+  };
+
+  class TupleExpr final : public Expr
+  {
+    std::vector<ExprPtr> elements;
+
+  public:
+    explicit TupleExpr(SourceLocation location, std::vector<ExprPtr> elements)
+    : Expr(Kind::Tuple, location), elements(std::move(elements))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Tuple;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("tuple").field(elements).finish();
+    }
+  };
+
+  class WhenExpr final : public Expr
+  {
+    std::vector<ExprPtr> arguments;
+    ExprPtr body;
+
+  public:
+    explicit WhenExpr(
+      SourceLocation location, std::vector<ExprPtr> arguments, ExprPtr body)
+    : Expr(Kind::When, location),
+      arguments(std::move(arguments)),
+      body(std::move(body))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::When;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("when").field(arguments).field(body).finish();
+    }
+  };
+
+  class NewExpr final : public Expr
+  {
+    std::vector<std::unique_ptr<MemberDef>> elements;
+    std::optional<Symbol> region;
+
+  public:
+    explicit NewExpr(
+      SourceLocation location,
+      std::vector<std::unique_ptr<MemberDef>> elements,
+      std::optional<Symbol> region)
+    : Expr(Kind::New, location), elements(std::move(elements)), region(region)
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::New;
+    }
+
+    // The implementation needs MemberDef to be complete, therefore it is moved
+    // to `print.cc`.
+    void print(NodePrinter& out) const override;
+  };
+
+  class LambdaExpr final : public Expr
```

Comment:
Does a lambda not have parameters?

Reply:
Yes they do, but they are part of a `signature` node in the untyped AST, which I've ignored for now.

I can add TODOs for fields that should exist in the future (I've done that in `conversion.cc` already).

Reply:
Yes please. I've been adding comments about them until I reached this comment. :)

Or better yet, add the fields directly even if you initialise them as null.